diff --git a/config/common.mk b/config/common.mk
index 56cd33d..e88b59e 100644
--- a/config/common.mk
+++ b/config/common.mk
@@ -186,31 +186,65 @@ PRODUCT_VERSION_MAINTENANCE = 0
 # Set CM_BUILDTYPE
 ifdef CM_NIGHTLY
     CM_BUILDTYPE := NIGHTLY
+    CM_EXTRAVERSION_DATESTAMP := TRUE
+    CM_EXTRAVERSION_TIMESTAMP_UTC := TRUE
+endif
+ifdef CM_SNAPSHOT
+    CM_BUILDTYPE := SNAPSHOT
+    CM_EXTRAVERSION_DATESTAMP := TRUE
+    CM_EXTRAVERSION_TIMESTAMP_UTC := TRUE
 endif
 ifdef CM_EXPERIMENTAL
     CM_BUILDTYPE := EXPERIMENTAL
+    CM_EXTRAVERSION_DATESTAMP := TRUE
+    CM_EXTRAVERSION_TIMESTAMP_UTC := TRUE
 endif
 ifdef CM_RELEASE
     CM_BUILDTYPE := RELEASE
+    CM_EXTRAVERSION := NOTUSED
 endif
 
-ifdef CM_BUILDTYPE
-    ifdef CM_EXTRAVERSION
-        # Force build type to EXPERIMENTAL
-        CM_BUILDTYPE := EXPERIMENTAL
-        # Add leading dash to CM_EXTRAVERSION
-        CM_EXTRAVERSION := -$(CM_EXTRAVERSION)
-    endif
-else
-    # If CM_BUILDTYPE is not defined, set to UNOFFICIAL
+# If CM_BUILDTYPE is not defined, set to UNOFFICIAL
+ifndef CM_BUILDTYPE
     CM_BUILDTYPE := UNOFFICIAL
-    CM_EXTRAVERSION :=
 endif
 
+ifndef CM_EXTRAVERSION
+    ifdef CM_EXTRAVERSION_TIMESTAMP
+        ifdef CM_EXTRAVERSION_TIMESTAMP_UTC
+            CM_EXTRAVERSION_TIMESTAMP := $(shell date -u +%Y%m%d_%H%M)
+        else
+            CM_EXTRAVERSION_TIMESTAMP := $(shell date +%Y%m%d_%H%M)
+        endif
+    else ifdef CM_EXTRAVERSION_DATESTAMP
+        ifdef CM_EXTRAVERSION_TIMESTAMP_UTC
+            CM_EXTRAVERSION_TIMESTAMP := $(shell date -u +%Y%m%d)
+        else
+            CM_EXTRAVERSION_TIMESTAMP := $(shell date +%Y%m%d)
+        endif
+    endif
+    ifdef CM_EXTRAVERSION_TAG
+        ifdef CM_EXTRAVERSION_TIMESTAMP
+            CM_EXTRAVERSION := $(CM_EXTRAVERSION_TIMESTAMP)-$(CM_EXTRAVERSION_TAG)
+        else
+            CM_EXTRAVERSION := $(CM_EXTRAVERSION_TAG)
+        endif
+    else
+        ifdef CM_EXTRAVERSION_TIMESTAMP
+            CM_EXTRAVERSION := $(CM_EXTRAVERSION_TIMESTAMP)
+        endif
+    endif
+endif
+
+
 ifdef CM_RELEASE
     CM_VERSION := $(PRODUCT_VERSION_MAJOR).$(PRODUCT_VERSION_MINOR).$(PRODUCT_VERSION_MAINTENANCE)$(PRODUCT_VERSION_DEVICE_SPECIFIC)-$(CM_BUILD)
 else
-    CM_VERSION := $(PRODUCT_VERSION_MAJOR)-$(shell date -u +%Y%m%d)-$(CM_BUILDTYPE)-$(CM_BUILD)$(CM_EXTRAVERSION)
+    ifdef CM_EXTRAVERSION
+        CM_VERSION := $(PRODUCT_VERSION_MAJOR)-$(CM_BUILDTYPE)-$(CM_EXTRAVERSION)-$(CM_BUILD)
+    else
+        CM_VERSION := $(PRODUCT_VERSION_MAJOR)-$(CM_BUILDTYPE)-$(CM_BUILD)
+    endif
 endif
 
 PRODUCT_PROPERTY_OVERRIDES += \
diff --git a/config/common_full_phone.mk b/config/common_full_phone.mk
index 135a23d..05cfef0 100644
--- a/config/common_full_phone.mk
+++ b/config/common_full_phone.mk
@@ -17,5 +17,5 @@ PRODUCT_COPY_FILES += \
 
 ifeq ($(TARGET_SCREEN_WIDTH) $(TARGET_SCREEN_HEIGHT),$(space))
     PRODUCT_COPY_FILES += \
-        vendor/cm/prebuilt/common/bootanimation/480.zip:system/media/bootanimation.zip
+        vendor/cm/prebuilt/common/bootanimation/768.zip:system/media/bootanimation.zip
 endif
