--- a/cdc_ncm.c	2012-09-20 14:43:14.982087333 +0200
+++ b/cdc_ncm.c	2012-09-20 14:44:06.170088857 +0200
@@ -1018,7 +1018,7 @@ static int cdc_ncm_rx_fixup(struct usbne
 			pr_debug("next packet should be on 8k boundery. Restart the rx queue with %d urb size\n", ctx->rx_max);
 			cdc_ncm_set_urb_size(dev, ctx->rx_max);
 		}
-		goto error;
+		goto done;
 	}
 
 	if (skb_in->len == CDC_NCM_MIN_TX_PKT && ctx->packet_cnt != 0) {
@@ -1115,6 +1115,7 @@ static int cdc_ncm_rx_fixup(struct usbne
 			usbnet_skb_return(dev, skb);
 		}
 	}
+done:
 	return 1;
 error:
 	return 0;
